### PR TITLE
Vacuum Forming Plastic Molds

### PR DESCRIPTION
**Pull request by Maleclypse**
#### Summary
Content "Vacuum Forming Plastic Molds"

#### Purpose of change
Partially addressing issues with plastic mold, by adding vacuum forming plastic molds.

#### Describe the solution
Adds vacuum former tools, recipe to make a tool, component to use tool, book to find recipe and add other casting processes to, itemgroups to spawn these things and edited plastic molding from tailoring requirements.  Not sure if this PR should obsolete plastic mold.  

#### Describe alternatives you've considered
None

#### Testing
TBD

### CODE REVIEW

File: data/json/items/book/fabrication.json [hunk 1]
```diff
@@ -43,6 +43,33 @@
...
   },
+  {
+    "id": "theater_props",
+    "type": "BOOK",
+    "looks_like": "glassblowing_book",
+    "name": {
+      "str": "The Prop Builder's: Molding and Casting Handbook",
+      "str_pl": "copies of The Prop Builder's: Molding and Casting Handbook"
+    },
+    "description": "The first book to contain, in one comprehensive volume every molding and casting procedure of use to the theater props builder, of every skill level.  The author demonstrates the techniques involved in over thirty materials from papier-mache to breakaway glass.",
```
Comment (anothersimulacrum) [inline data/json/items/book/fabrication.json:54]:
```suggestion
    "description": "The first book to contain, in one comprehensive volume, every molding and casting procedure of use to the theater props builder of any skill level.  The author demonstrates the techniques involved in over thirty materials from papier-mache to breakaway glass.",
```

File: data/json/items/book/fabrication.json [hunk 2]
```diff
@@ -43,6 +43,33 @@
     "intelligence": 8,
     "time": "30 m"
   },
+  {
+    "id": "theater_props",
+    "type": "BOOK",
+    "looks_like": "glassblowing_book",
+    "name": {
+      "str": "The Prop Builder's: Molding and Casting Handbook",
```
Comment (chaohedgehogs) [inline data/json/items/book/fabrication.json:51]:
```suggestion
      "str": "The Prop Builder's Molding and Casting Handbook",
```

File: data/json/items/book/fabrication.json [hunk 3]
```diff
@@ -43,6 +43,33 @@
...
+      "str_pl": "copies of The Prop Builder's: Molding and Casting Handbook"
```
Comment (chaohedgehogs) [inline data/json/items/book/fabrication.json:52]:
```suggestion
      "str_pl": "copies of The Prop Builder's Molding and Casting Handbook"
```

File: data/json/items/resources/plastic.json [hunk 4]
```diff
@@ -60,6 +60,21 @@
     "color": "light_blue",
     "use_action": { "type": "deploy_furn", "furn_type": "f_plastic_groundsheet" }
   },
+  {
+    "id": "polycarbonate_sheet",
+    "type": "GENERIC",
+    "category": "spare_parts",
+    "name": { "str": "polycarbonate sheet" },
+    "description": "This is a stack of polycarbonate sheets that can be used for vacuum molding.",
```
Comment (actual-nh) [inline data/json/items/resources/plastic.json:68]:
```suggestion
    "description": "This is a stack of polycarbonate sheets that can be used for, among other things, vacuum molding.",
```

**Comment by Maleclypse** [comment]
```TEST_CASE( "tools use charge to craft", "[crafting][charge]" )
{
    std::vector<item> tools;

    GIVEN( "recipe and required tools/materials" ) {
        recipe_id carver( "carver_off" );
        // Uses fabrication skill
        // Requires electronics 3
        // Difficulty 4
        // Learned from advanced_electronics or textbook_electronics

        // Tools needed:
        tools.emplace_back( "screwdriver" );
        tools.emplace_back( "mold_plastic" );

        // Materials needed
        tools.insert( tools.end(), 10, item( "solder_wire" ) );
        tools.insert( tools.end(), 6, item( "plastic_chunk" ) );
        tools.insert( tools.end(), 2, item( "blade" ) );
        tools.insert( tools.end(), 5, item( "cable" ) );
        tools.emplace_back( "motor_tiny" );
        tools.emplace_back( "power_supply" );
        tools.emplace_back( "scrap" );

        // Charges needed to craft:
        // - 10 charges of soldering iron
        // - 10 charges of surface heat

        WHEN( "each tool has enough charges" ) {
            item hotplate = tool_with_ammo( "hotplate", 20 );
            REQUIRE( hotplate.ammo_remaining() == 20 );
            tools.push_back( hotplate );
            item soldering = tool_with_ammo( "soldering_iron", 20 );
            REQUIRE( soldering.ammo_remaining() == 20 );
            tools.push_back( soldering );

            THEN( "crafting succeeds, and uses charges from each tool" ) {
                prep_craft( recipe_id( "carver_off" ), tools, true );
                int turns = actually_test_craft( recipe_id( "carver_off" ), INT_MAX );
                CAPTURE( turns );
                CHECK( get_remaining_charges( "hotplate" ) == 10 );
                CHECK( get_remaining_charges( "soldering_iron" ) == 10 );
            }
        }```
OK so here is the test I've broken.   @wapcaplet  Can you walk me through how to fix one of these?  I think I'd replace the plastic mold with the makeshift vacuum mold and then add 2 polycarbonate sheets in materials needed?  Is there anything else I'd be missing?

**Comment by wapcaplet** [comment]
@Maleclypse Sorry, I am afraid I was never able to comprehend the crafting tests. I can reproduce the failures (3) in your branch when running them locally, but I am just as baffled as you are about why they are failing.

File: data/json/items/tool/workshop.json [hunk 5]
```diff
@@ -130,6 +130,62 @@
...
+        "max_contains_weight": "20 kg",
+        "item_restriction": [ "medium_battery_cell", "medium_plus_battery_cell", "medium_atomic_battery_cell", "medium_disposable_cell" ]
+      }
+    ]
+  },
+  {
+    "id": "vac_mold",
+    "type": "TOOL",
+    "name": { "str": "vacuum molder" },
+    "description": "This is a portable vacuum mold unit with an air pump.  It's used for creating plastic molds.",
```
Comment (Maleclypse) [inline data/json/items/tool/workshop.json:165]:
```suggestion
    "description": "This is a portable vacuum forming machine with an air pump.  It's used for creating plastic molds.",
```

File: data/json/items/tool/workshop.json [hunk 6]
```diff
@@ -130,6 +130,62 @@
     "symbol": ";",
     "color": "brown"
   },
+  {
+    "id": "makeshift_vacuum_mold",
+    "type": "TOOL",
+    "name": { "str": "makeshift vacuum mold" },
+    "description": "This is a homemade vacuum molder unit with an air pump.  It's used for creating plastic molds.",
```
Comment (Maleclypse) [inline data/json/items/tool/workshop.json:137]:
```suggestion
    "description": "This is a homemade vacuum forming machine with an air pump.  It's used for creating plastic molds.",
```

**Comment by Maleclypse** [comment]
@anothersimulacrum I am lost on what I need to do to fix these tests.

**Comment by anothersimulacrum** [comment]
The failure at `crafting succeeds, and uses charges from each tool` is due to missing qualities.
The failure at `crafting succeeds, and uses charges from multiple tools` is due to missing qualities and tools.
The failure at `crafting succeeds, and uses charges from the UPS` is due to missing qualities and tools.

This obviously isn't enough for you to go on, I just want to note it down now so I don't forget.

**Comment by Maleclypse** [comment]
@anothersimulacrum Would it be best for me to close out my PR's targeting 0.F and reopen them targeting master?

**Comment by anothersimulacrum** [comment]
@Maleclypse If you can do an interactive rebase, it would be best to rebase them onto master, and drop all the commits that are not yours, then force push back.

**Comment by Maleclypse** [comment]
I think I picked the right conflicts but some of them I'm not sure.  Please review to make sure I haven't jacked anything up.

File: data/json/furniture_and_terrain/terrain-railroads.json [hunk 7]
```diff
@@ -13,8 +13,7 @@
       "ter_set": "t_null",
       "sound": "crunch!",
       "sound_fail": "whump!",
-      "items": [ { "item": "material_gravel", "count": [ 5, 15 ] }, { "item": "sharp_rock", "count": [ 0, 1 ] } ],
-      "bash_below": true
+      "items": [ { "item": "material_gravel", "count": [ 5, 15 ] }, { "item": "sharp_rock", "count": [ 0, 1 ] } ]
```
Comment (Maleclypse) [inline data/json/furniture_and_terrain/terrain-railroads.json:16]: I don't want this change

File: data/json/recipes/recipe_obsolete.json [hunk 8]
```diff
@@ -2775,7 +2775,7 @@
   },
   {
     "type": "recipe",
-    "result": "rifle_223",
+    "result": "bodypillow",
```
Comment (Maleclypse) [inline data/json/recipes/recipe_obsolete.json:2778]: This isn't my change.

File: src/requirements.cpp [hunk 9]
```diff
@@ -745,17 +745,21 @@ bool requirement_data::has_comps( const read_only_visitable &crafting_inv,
 {
     bool retval = true;
     int total_UPS_charges_used = 0;
-    for( const auto &set_of_tools : vec ) {
```
Comment (Maleclypse) [inline src/requirements.cpp:748]: I don't think this file is my changes but it might be anothersimulacrum when they fixed the tests failing for me.

**Comment by phaethonfire** [comment]
Not sure if this is better as a bug report, but this book is probably going to have the same problem as the old plastics book in that it won't spawn in libraries.  Libraries are using a static set of books on the shelves that completely ignore these item groups.  So newly added books, like "theater_props", which would quite reasonable spawn in libraries and DO spawn in school libraries don't actually appear in libraries.

I know this PR was closed, but I am pretty sure that the "theater_props" book needs to be incorporated in this file as well:
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/data/json/itemgroups/SUS/library.json

Has anyone else tested this?  I discovered the issue coming off the back of a 0.F stable playthrough.

**Comment by Maleclypse** [comment]
I hear you and will have a PR addressing this tonight.

**Pull request merged**